text/1006-groupped-cli.rst

```diff
@@ -0,0 +1,240 @@
+all seems very limiting. Therefore we propose the following:
+
+* Inspect commands will be exposed **to the CLI** via the ``edgedb inspect``
+  command group.
+
+* Inspect subcommands will be exposed **in REPL** directly, without
+  requiring users to type ``inspect``. E.g. instead of typing
+  ``\inspect list-object-types`` users should be able to continue
+  using the ``\list-object-types`` command.
+
+Overall, the consistency of the CLI is more important than the minor
+inconsistencies of the introspection commands in REPL.
+
+
+Occasional Duplication of Commands
+----------------------------------
+
+Some of the commands will have aliases:
+
+* ``edgedb migrate`` is an alias for ``edgedb migration apply``. The reason
+  for having the alias: this will be a very popular and frequently typed
+  command.
+
+* ``edgedb inspect list-X`` might become aliases for ``edgedb X list``
+  for some types of entities in the future.  While this does not seem like
+  a "pure" solution, there is no harm in having aliases like this.
+
+In general, we believe that having aliases for some commands cannot
+harm the overall developer experience of using the CLI.
+
+
+RFC 1003 -- Rejected Ideas
+--------------------------
+
+The superseded RFC 1003 explicitly rejected the ``<group> <command>`` naming
+scheme, quote::
+
+    * inability to adjust every command naturally in this way;
+
+    * disruptive nature of the change;
+
+    * less verbose ``help`` output; and
+
+    * less natural-sounding commands.
+
+The verbosity of the updated ``edgedb --help`` output can be and will be
+tweaked until it hits the perfect balance of being readable and informative.
+
+The less natural-sounding commands argument is valid, as
+``edgedb create-migration`` certainly sounds more natural than
+``edgedb migration create``. But given that we will likely have between more
+than 30 subcommands, it is clear that giving users a way to organize
+subcommands mentally in categories to memorize the overall structure is more
+important than "making commands sound like plain English".
+
+The proposed change is indeed very disruptive but we believe it is still worth
+implementing it before 1.0. It is important to understand that RFC 1003 was
+written when the CLI had only one "tool within a tool" — ``edgedb server``.
+Since then we have added ``edgedb project`` and it became apparent that we
+will likely continue to add more tools like that.
+
+
+Backwards Compatibility
+=======================
+
+We will supporting all existing commands (e.g. ``edgedb create-migration``)
+until the 1.0 release.
+
+The old commands will be hidden from the ``--help`` output. When run, old
+commands will render a deprecation warning, e.g.::
+
+    $ edgedb create-migration
+    The `create-migration` command has been deprecated.
+    Use `edgedb migration create` instead.
+
+
+Open Questions
+==============
+
+* Do we need to remove or redesign ``edgedb configure``?
+
+* Do we need ``edgedb query``?
```

Comment:
Sure. What's wrong with `edgedb query`?

Reply:
Well, `-c` used in command which don't have subcommands. I.e. `sh -c` is there because `sh exec` would search for a file named `exec`. Same with `python`, etc.

But we have subcommands, so some subcommand makes more sense. Also the subcommand can have additional nice arguments like `edgedb query -f file.esdl` or `edgedb query --format=json` or `edgedb query --show-times`, which would be ugly to have as global arguments (well, we already sorta have format, but I'm not sure this is a great idea).